### PR TITLE
rpc, wallet: Scan mempool after import* - Second attempt

### LINKED ISSUE #18954: import* wallet RPCs don't "rescan" the mempool
Steps to reproduce:

* Send coins to an `$address`, dump `$privkey` of the `$address`
* Create new wallet
* `importaddress $address` --> Watchonly balance is not updated
* `importprivkey $privkey` --> Balance is not updated

Needs a restart to get fixed. See also #11477

### PR DESCRIPTION
**Pull request by fjahr**
This PR picks up the work from #18964 and closes #18954.

It should incorporate all the unaddressed feedback from the PR:
- Mempool rescan now expanded to all relevant import* RPCs
- Added documentation in the help of each RPC
- More tests

### CODE REVIEW

**Comment by achow101** [comment]
ACK abdaa251610569a346af55a1121c8ab385efb655

**Review by Sjors — COMMENTED** [review]
I'm unable to get this to work with descriptor wallets; still seeing the issue I reported in #25453.

Step 1: create a regular descriptor wallet, put some coins in it (I used a Taproot address for this)
Step 2: get ~one of its~ the matching receive descriptor ~s~ via `listdescriptors`
Step 3: create a watch-only wallet
Step 4: send coins to self, and before it confirms do:
Step 5: call `importdescriptors` on the watch-only wallet, using the above descriptor and some recent timestamp.

It will not add the mempool transaction until you either restart or it confirms.

`rescanblockchain` also doesn't add the transaction, but I suppose that's out of scope here.

_Update:_ Note that you DO see the confirmed deposit transaction, but you don't see the outgoing mempool transaction.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1797,6 +1798,10 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
             }
         }
     }
+    if (!max_height) {
+        WalletLogPrintf("Scanning current mempool transactions.\n");
```
Comment (Sjors) [inline src/wallet/wallet.cpp:1802]: I don't see this notice in the log either.

**Comment by w0xlt** [comment]
@Sjors I was able to retrieve the mempool transactions by following these steps on regtest and signet.

**Comment by Sjors** [comment]
> @Sjors I was able to retrieve the mempool transactions by following these steps on regtest and signet.

Mmm, and without this PR you're able to reproduce the original issue?

File: test/functional/wallet_importdescriptors.py [hunk 2]
```diff
@@ -480,7 +480,7 @@ def run_test(self):
         addr = wmulti_pub.getnewaddress('', 'bech32')
         assert_equal(addr, 'bcrt1qp8s25ckjl7gr6x2q3dx3tn2pytwp05upkjztk6ey857tt50r5aeqn6mvr9') # Derived at m/84'/0'/0'/1
         change_addr = wmulti_pub.getrawchangeaddress('bech32')
-        assert_equal(change_addr, 'bcrt1qt9uhe3a9hnq7vajl7a094z4s3crm9ttf8zw3f5v9gr2nyd7e3lnsy44n8e')
+        assert_equal(change_addr, 'bcrt1qzxl0qz2t88kljdnkzg4n4gapr6kte26390gttrg79x66nt4p04fssj53nl')
```
Comment (w0xlt) [inline test/functional/wallet_importdescriptors.py:483]: Perhaps the assertions below can be added.
They test that `importdescriptors` retrieved the mempool transaction.

Not related to this PR, but transaction `send_txid` is confirmed in line 457, isn't it ? 
Why is it still in the mempool in line 463 and 470?

```suggestion
        assert_equal(change_addr, 'bcrt1qzxl0qz2t88kljdnkzg4n4gapr6kte26390gttrg79x66nt4p04fssj53nl')
        assert(any([x == send_txid for x in self.nodes[0].getrawmempool(True)]))
        assert(any([x['txid'] == send_txid for x in wmulti_pub.listunspent(0)]))
```

File: src/wallet/test/wallet_tests.cpp [hunk 3]
```diff
@@ -765,7 +765,7 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
     // being blocked
     wallet = TestLoadWallet(context);
     BOOST_CHECK(rescan_completed);
-    BOOST_CHECK_EQUAL(addtx_count, 2);
+    BOOST_CHECK_EQUAL(addtx_count, 3);
```
[the remark this answers is not in the document]
Comment (w0xlt) [reply src/wallet/test/wallet_tests.cpp:768]: Agreed.

**Review by w0xlt — APPROVED** [review]
Tested ACK https://github.com/bitcoin/bitcoin/pull/25351/commits/abdaa251610569a346af55a1121c8ab385efb655

Some nits:

**Comment by w0xlt** [comment]
@Sjors Yes, without this PR (using the master branch), I am able to reproduce the issue.

**Comment by Sjors** [comment]
@Sjors and your second wallet was a watch-only wallet? Guess I'll have to retry to check my own sanity :-)

**Comment by w0xlt** [comment]
@Sjors the watch-only was created with `disable_private_keys=true blank=true` and the `tr(xpub...)` descriptors (internal and external) were imported from the first wallet.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (fjahr) [reply src/wallet/wallet.cpp:1802]: Strange, that may be a hint that the build didn't have this change? But I will go through your steps as well as an additional check.

File: test/functional/wallet_importdescriptors.py (hunk 2 above)
Comment (fjahr) [reply test/functional/wallet_importdescriptors.py:483]: Added

> Not related to this PR, but transaction `send_txid` is confirmed in line 457, isn't it ? Why is it still in the mempool in line 463 and 470?

Hm, you are right that this shouldn't really work but it still does due to a race between the two nodes involved. I am kind of surprised that this doesn't lead flakiness in the CI but I couldn't find any related issue. I addressed this in #25476.

File: src/wallet/rpc/backup.cpp [hunk 4]
```diff
@@ -1592,7 +1600,7 @@ RPCHelpMan importdescriptors()
         "                                                              Use the string \"now\" to substitute the current synced blockchain time.\n"
         "                                                              \"now\" can be specified to bypass scanning, for outputs which are known to never have been used, and\n"
         "                                                              0 can be specified to scan the entire blockchain. Blocks up to 2 hours before the earliest timestamp\n"
-        "                                                              of all descriptors being imported will be scanned.",
+        "                                                              of all descriptors being imported will be scanned as well as the mempool.",
```
[the remark this answers is not in the document]
Comment (fjahr) [reply src/wallet/rpc/backup.cpp]: done

File: test/functional/wallet_import_rescan.py [hunk 5]
```diff
@@ -223,6 +230,46 @@ def run_test(self):
...
+                label=variant.label,
+                address_type=variant.address_type.value,
+            ))
+            variant.key = self.nodes[1].dumpprivkey(variant.address["address"])
+            variant.initial_amount = get_rand_amount()
+            variant.initial_txid = self.nodes[0].sendtoaddress(variant.address["address"], variant.initial_amount)
+            variant.confirmation_height = 0
+            variant.timestamp = timestamp
+
+        assert_equal(len(self.nodes[0].getrawmempool()), 54)
```
[the remark this answers is not in the document]
Comment (fjahr) [reply test/functional/wallet_import_rescan.py]: done

File: src/wallet/test/wallet_tests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (fjahr) [reply src/wallet/test/wallet_tests.cpp]: Yeah, I see what you mean. The `TestLoadWallet()` helper function had a call to `postInitProcess()` which meant that the mempool was checked twice. Removing that led to the numbers in the `CreateWallet` test being much more sensible and there seem to be no other negative effects from that change. At least the test isn't failing and I also didn't see any other reason to keep it. I added some comments as well.

**Comment by fjahr** [comment]
Addressed feedback but have yet to go through @Sjors steps to see if I can reproduce his issue.

File: src/wallet/test/wallet_tests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/wallet/test/wallet_tests.cpp]: IMO bumping the numbers was better. :/

File: test/functional/wallet_importdescriptors.py [hunk 6]
```diff
@@ -480,7 +480,9 @@ def run_test(self):
...
+        assert(any([x == send_txid for x in self.nodes[0].getrawmempool(True)]))
+        assert(any([x['txid'] == send_txid for x in wmulti_pub.listunspent(0)]))
```
Comment (luke-jr) [inline test/functional/wallet_importdescriptors.py:485]:
```suggestion
        assert(any(x == send_txid for x in self.nodes[0].getrawmempool(True)))
        assert(any(x['txid'] == send_txid for x in wmulti_pub.listunspent(0)))
```

No benefit to making it a list - just causes every element to be compared (using the generator directly stops comparing as soon as it finds a match)

Or even better and clearer...

```suggestion
        assert(send_txid in self.nodes[0].getrawmempool(True))
        assert(send_txid in (x['txid'] for x in wmulti_pub.listunspent(0)))
```

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (Sjors) [reply src/wallet/wallet.cpp:1802]: Indeed it turns out I was running the wrong build (again). Let me try again. At least I know for sure the bug exists on master 506d9b25a3f8ef8a4ce63f989083803d8d4ae6df :-)

**Review by Sjors — APPROVED** [review]
tACK 8f59d58549ab8a9e7ecf8eb3ed2fb6d18b812a6d

(though I only manually tested with a taproot descriptor)

File: src/wallet/test/wallet_tests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (fjahr) [reply src/wallet/test/wallet_tests.cpp]: > IMO bumping the numbers was better. :/

Could you say why you like it better? When the numbers are bumped I feel they don't match what the test is doing and I would need to add more comments to explain why this is the case. Hence it seemed like the better choice to me. Also curious to hear your feedback here @w0xlt and @Sjors :)

File: test/functional/wallet_importdescriptors.py (hunk 6 above)
Comment (fjahr) [reply test/functional/wallet_importdescriptors.py]: done

**Comment by fjahr** [comment]
Addressed feedback from @luke-jr

**Review by w0xlt — APPROVED** [review]
reACK https://github.com/bitcoin/bitcoin/pull/25351/commits/1be796418934ae7370cb0ed501877db59e738106

File: src/wallet/test/wallet_tests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/wallet/test/wallet_tests.cpp]: Seems like it would be undefined behaviour (between modules) to skip `postInitProcess`.

Counting AddToWallet calls might be too low-level to be a good test, but that seems out of scope here IMO.

**Comment by Sjors** [comment]
re-utACK 1be796418934ae7370cb0ed501877db59e738106 (only a test change)

File: src/wallet/test/wallet_tests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Sjors) [reply src/wallet/test/wallet_tests.cpp]: I have no strong feelings about this, as long as it's clear from the code documentation where the transactions are coming from.

**Comment by achow101** [comment]
ACK 1be796418934ae7370cb0ed501877db59e738106
